src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/AadPlugin.cs

```diff
@@ -27,25 +30,80 @@ public AadPlugin(IWamProxy wamProxy, IWebAccountProviderFactory webAccountProvid
+        /// </summary>
+        public async Task<IReadOnlyList<IAccount>> GetAccountsAsync(
+            string clientId,
+            string authority,
+            Cache.ICacheSessionManager cacheSessionManager,
+            Instance.Discovery.IInstanceDiscoveryManager instanceDiscoveryManager)
         {
             var webAccountProvider = await _webAccountProviderFactory.GetAccountProviderAsync("organizations").ConfigureAwait(false);
+            var wamAccounts = await _wamProxy.FindAllWebAccountsAsync(webAccountProvider, clientId).ConfigureAwait(false);
 
-            var webAccounts = await _wamProxy.FindAllWebAccountsAsync(webAccountProvider, clientID).ConfigureAwait(false);
-
-            var msalAccounts = webAccounts
-                .Select(webAcc => ConvertToMsalAccountOrNull(webAcc))
-                .Where(a => a != null)
-                .ToList();
+            if (wamAccounts.Count > 0)
+            {
+                var webAccountEnvs = wamAccounts
+                    .Select(w =>
+                    {
+                        _wamProxy.TryGetAccountProperty(w, "Authority", out string accountAuthority);
+                        if (accountAuthority != null)
+                        {
+                            return (new Uri(accountAuthority)).Host;
+                        }
+                        else
+                        {
+                            _logger.WarningPii(
+                                $"[WAM AAD Provider] Could not convert the WAM account {w.UserName} (id: {w.Id}) to an MSAL account because the Authority could not be found",
+                                $"[WAM AAD Provider] Could not convert the WAM account {w.Id} to an MSAL account because the Authority could not be found");
+
+                            return null;
+                        }
+                    })
+                    .Where(a => a != null);
+
+                var instanceMetadata = await instanceDiscoveryManager.GetMetadataEntryTryAvoidNetworkAsync(authority, webAccountEnvs, cacheSessionManager.RequestContext)
+                    .ConfigureAwait(false);
+                var accountsFromCache = await cacheSessionManager.GetAccountsAsync().ConfigureAwait(false);
+
+                var msalAccountTasks = wamAccounts
+                    .Select(
+                        async webAcc =>
+                            await ConvertToMsalAccountOrNullAsync(
+                                clientId,
+                                webAcc,
+                                instanceMetadata,
+                                cacheSessionManager,
+                                accountsFromCache).ConfigureAwait(false));
+
+                var msalAccounts = (await Task.WhenAll(msalAccountTasks).ConfigureAwait(false)).Where(a => a != null).ToList();
+
+                _logger.Info($"[WAM AAD Provider] GetAccountsAsync converted {wamAccounts.Count()} MSAL accounts");
```

Comment:
assuming all the wamAccounts were converted...would it be helpful to know how many msalAccounts were created?